src/ooni/ooni_test.cpp

```diff
@@ -0,0 +1,159 @@
+                    logger->warn("geoip files not configured; skipping");
+                } else {
+                    ErrorOr<nlohmann::json> res = geoip(ip, country_p, asn_p);
+                    if (!!res) {
+                        logger->debug("GeoIP result: %s", res->dump().c_str());
+                        // Since `geoip()` sets defaults before querying, the
+                        // following accesses of json should not fail unless for
+                        // programmer error after refactoring. In that case,
+                        // better to let the exception unwind than just print
+                        // a warning, because the former is easier to notice
+                        // and therefore fix during development
+                        probe_asn = (*res)["asn"];
+                        logger->info("probe_asn: %s", probe_asn.c_str());
+                        probe_cc = (*res)["country_code"];
+                        logger->info("probe_cc: %s", probe_cc.c_str());
+                    }
+                }
+            }
+            cb();
+        },
+        options, reactor, logger);
+}
+
+void OoniTest::open_report() {
+    if (output_filepath == "") {
+        output_filepath = generate_output_file_path();
+    }
+    file_report.test_name = test_name;
+    file_report.test_version = test_version;
+    file_report.test_start_time = test_start_time;
+
+    file_report.options = options;
+    file_report.filename = output_filepath;
+
+    file_report.probe_ip = probe_ip;
+    file_report.probe_cc = probe_cc;
+    file_report.probe_asn = probe_asn;
+
+    file_report.open();
+}
+
+std::string OoniTest::generate_output_file_path() {
+    std::string filename;
+    char buffer[100];
+    strftime(buffer, sizeof(buffer), "%FT%H%M%SZ", &test_start_time);
+    std::string iso_8601_date(buffer);
+    filename = "report-" + test_name + "-";
+    filename += iso_8601_date + ".json";
+    return filename;
+}
+
+void OoniTest::begin(Callback<> cb) {
+    mk::utc_time_now(&test_start_time);
+    geoip_lookup([=]() {
+        open_report();
+        if (needs_input) {
+            if (input_filepath == "") {
+                // TODO: allow to pass error to cb()
+                logger->warn("an input file is required");
+                cb();
+                return;
+            }
+            input_generator.reset(new std::ifstream(input_filepath));
+            if (!input_generator->good()) {
+                logger->warn("cannot read input file");
+                cb();
+                return;
+            }
+        } else {
+            input_generator.reset(new std::istringstream("\n"));
+        }
+        run_next_measurement(cb);
+    });
+}
+
+void OoniTest::end(Callback<> cb) {
+    file_report.close();
+    collector::submit_report(
+        output_filepath,
+        options.get("collector_base_url", collector::default_collector_url()),
+        [=](Error) { cb(); }, options, reactor, logger);
+}
+
```

Comment:
No changes here, only taken verbatim from `src/ooni/ooni_test_impl.hpp` and reformatted